### PR TITLE
Feat: added authentication

### PR DESCRIPTION
**Pull request by Chakravarthi0**
Implemented authentication using firebase. User can create an account and login using their account.

Added the following components:
Loader
Loading-error
Overlay

Added the following page:
User profile page

### CODE REVIEW

File: src/utils/fireBaseHelper.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  if (!snapShot.exists()) {
+    const { firstName, lastName } = userData;
+    try {
+      await setDoc(currentUserRef, {
+        email: user.email,
+        firstName,
+        lastName,
+      });
+    } catch (err) {
+      console.log(err);
```
Comment (manojasarsa) [inline src/utils/fireBaseHelper.ts:21]: You can make use of `console.error` here to handle error.

**Review by manojasarsa — APPROVED** [review]
Code looks clean. GTG!

File: src/app-routes/AppRoutes.tsx [hunk 2]
```diff
@@ -8,19 +9,40 @@ import {
...
 } from "../pages";
+import { authContext } from "../context";
+import ProtectedRoutes from "./ProtectedRoutes";
 
 function AppRoutes() {
+  const {
+    authState: {
+      userDetails: { token },
+    },
+  } = useContext(authContext);
```
Comment (msdeshmukh009) [inline src/app-routes/AppRoutes.tsx]: Maybe you can use the useAuth hook here which you have already made.

File: src/hooks/useAuth.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    password: string;
+    firstName: string;
+    lastName: string;
+  };
+  const signUp = async ({
+    userEmail,
+    password,
+    firstName,
+    lastName,
+  }: signUpType) => {
```
Comment (msdeshmukh009) [inline src/hooks/useAuth.ts]: I think it's better to either extract this type or directly mention it in the brackets themselves.
```js
const signUp = async ({
    userEmail,
    password,
    firstName,
    lastName,
  }:{
    userEmail: string;
    password: string;
    firstName: string;
    lastName: string;
  } )
  ```

**Review by msdeshmukh009 — APPROVED** [review]
The functionality is working as expected.
The code is easily readable and well written.
Nice work!!

File: src/app-routes/AppRoutes.tsx (hunk 2 above)
Comment (Chakravarthi0) [reply src/app-routes/AppRoutes.tsx]: Thanks for suggestions, I have updated the code.

**Pull request merged**